05-lists-and-dictionaries/poetry.py

```diff
@@ -0,0 +1,102 @@
+# Copyright (c) 2016 Ami . All rights reserved
+
+from random import choice
+import re
+
+class SpeechPart(object):
+    def __init__(self, name, count, vocabulary):
+        self.name = name
+        self.count = count
+        self.vocabulary = vocabulary
+
+    def random_words(self):
+        words = []
+
+        while True:
+            if len(words) == self.count or len(words) == len(self.vocabulary):
+                return words
+            else:
+                selected_word = choice(self.vocabulary)
+                if selected_word not in words:
+                    words.append(selected_word)
+
+        return words
+
+def make_poem():
+    speech_parts = [
+        SpeechPart(
+            name="verb",
+            count=3,
+            vocabulary=["kicks", "jingles", "bounces", "slurps", "meows", "explodes", "curdles"]
+        ),
+        SpeechPart(
+            name="adjective",
+            count=3,
+            vocabulary=["furry", "balding", "incredulous", "fragrant", "exuberant", "glistening"]
+        ),
+        SpeechPart(
+            name="adverb",
+            count=1,
+            vocabulary=["curiously", "extravagantly", "tantalizingly", "furiously", "sensuously"]
+        ),
+        SpeechPart(
+            name="noun",
+            count=3,
+            vocabulary=[
+                "fossil",
+                "horse",
+                "aardvark",
+                "judge",
+                "chef",
+                "mango",
+                "extrovert",
+                "gorilla"
+            ]
+        ),
+        SpeechPart(
+            name="preposition",
+            count=2,
+            vocabulary=[
+                "against",
+                "after",
+                "into",
+                "beneath",
+                "upon",
+                "for",
+                "in",
+                "like",
+                "over",
+                "within"
+            ]
+        )
+    ]
+
+    poem_structure = (
+        "{A/An} {adjective1} {noun1}\n\n"
+        "{A/An} {adjective1} {noun1} {verb1} {preposition1} the {adjective2} {noun2}\n"
+        "{adverb1}, the {noun1} {verb2}\n"
+        "the {noun2} {verb3} {preposition2} a {adjective3} {noun3}\n"
+    )
+
+    transform_dict = {}
+
+    for part in speech_parts:
+        selected_words = part.random_words()
+
+        if part.name == "adjective":
+            pattern = re.compile('^(a|e|i|a|o|u)')
```

Comment:
Please extract a constant variable for this pattern string with some kind of meaningful name (e.g. FIRST_LETTER_VOWEL).
Note that you're still compiling this regex inside the loop, but it doesn't really matter for this exercise since it will happen only one time because there's only 1 'adjective' speech part.